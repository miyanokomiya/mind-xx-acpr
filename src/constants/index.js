export const INTERVAL_CLICK = 300
export const INTERVAL_DOUBLE_CLICK = 300

export const NODE_MARGIN_X = 20
export const NODE_MARGIN_Y = 20
