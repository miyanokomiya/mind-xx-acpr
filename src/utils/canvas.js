export const getPoint = e => {
  let x, y
  let rect = e.currentTarget.getBoundingClientRect()
  let dx = rect.left + window.pageXOffset
  let dy = rect.top + window.pageYOffset

  if (isTouch(e)) {
    let touch = e.touches[0]
    x = touch.pageX - dx
    y = touch.pageY - dy
  } else {
    x = e.pageX - dx
    y = e.pageY - dy
  }
  return {
    x: x,
    y: y
  }
}

export const isTouch = e => {
  return e.touches
}

export const isMulitTouch = e => {
  return isTouch(e) && e.touches.length > 1
}

export const isTouchExist = e => {
  return isTouch(e) && e.touches.length > 0
}

export const getPoints = e => {
  let ret = []
  let rect = e.currentTarget.getBoundingClientRect()
  let dx = rect.left + window.pageXOffset
  let dy = rect.top + window.pageYOffset

  for (let i = 0; i < e.touches.length; i++) {
    let touch = e.touches[i]
    ret.push({
      x: touch.pageX - dx,
      y: touch.pageY - dy
    })
  }

  return ret
}
