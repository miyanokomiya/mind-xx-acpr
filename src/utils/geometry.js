export function getCoveredRectangle ({ positions, sizes }) {
  const keys = Object.keys(positions).filter(key => {
    return positions[key] && sizes[key]
  })
  if (keys.length === 0) {
    return {
      x: 0,
      y: 0,
      width: 0,
      height: 0
    }
  }
  let recP = keys.reduce(
    (p, key) => {
      return {
        x: Math.min(p.x, positions[key].x),
        y: Math.min(p.y, positions[key].y)
      }
    },
    {
      x: positions[keys[0]].x,
      y: positions[keys[0]].y
    }
  )
  const recS = keys.reduce(
    (p, key) => {
      return {
        width: Math.max(p.width, positions[key].x + sizes[key].width - recP.x),
        height: Math.max(
          p.height,
          positions[key].y + sizes[key].height - recP.y
        )
      }
    },
    {
      width: sizes[keys[0]].width,
      height: sizes[keys[0]].height
    }
  )
  return {
    x: recP.x,
    y: recP.y,
    width: recS.width,
    height: recS.height
  }
}

export function getCenterOfRectangle (rectangle) {
  return {
    x: rectangle.x + rectangle.width / 2,
    y: rectangle.y + rectangle.height / 2
  }
}

export function getDistance (p1, p2) {
  return Math.pow(Math.pow(p1.x - p2.x, 2) + Math.pow(p1.y - p2.y, 2), 1 / 2)
}

export function getClosestRectangle ({ rectangles, target }) {
  const keys = Object.keys(rectangles)
  if (keys.length === 0) {
    return null
  }
  let ret = null
  let minD = Number.POSITIVE_INFINITY
  const targetCenter = getCenterOfRectangle(target)
  keys.forEach(key => {
    const center = getCenterOfRectangle(rectangles[key])
    const d = getDistance(center, targetCenter)
    if (d < minD) {
      ret = key
      minD = d
    }
  })
  return ret
}

export function getRelationBetweenRectangleAndPoint ({ rectangle, point }) {
  if (point.x < rectangle.x) {
    if (point.y < rectangle.y) {
      return 1
    } else if (point.y > rectangle.y + rectangle.height) {
      return 7
    } else {
      return 4
    }
  } else if (point.x > rectangle.x + rectangle.width) {
    if (point.y < rectangle.y) {
      return 3
    } else if (point.y > rectangle.y + rectangle.height) {
      return 9
    } else {
      return 6
    }
  } else {
    if (point.y < rectangle.y) {
      return 2
    } else if (point.y > rectangle.y + rectangle.height) {
      return 8
    } else {
      return 5
    }
  }
}

export function getDistanceBetweenRectangleAndPoint ({ rectangle, point }) {
  const relation = getRelationBetweenRectangleAndPoint({ rectangle, point })
  switch (relation) {
    case 1:
      return getDistance(point, { x: rectangle.x, y: rectangle.y })
    case 2:
      return rectangle.y - point.y
    case 3:
      return getDistance(point, {
        x: rectangle.x + rectangle.width,
        y: rectangle.y
      })
    case 4:
      return rectangle.x - point.x
    case 5:
      return 0
    case 6:
      return point.x - (rectangle.x + rectangle.width)
    case 7:
      return getDistance(point, {
        x: rectangle.x,
        y: rectangle.y + rectangle.height
      })
    case 8:
      return point.y - (rectangle.y + rectangle.height)
    case 9:
      return getDistance(point, {
        x: rectangle.x + rectangle.width,
        y: rectangle.y + rectangle.height
      })
  }
}

export function getClosestRectangleByPoint ({ rectangles, point }) {
  const keys = Object.keys(rectangles)
  if (keys.length === 0) {
    return null
  }
  let ret = null
  let minD = Number.POSITIVE_INFINITY
  keys.forEach(key => {
    const d = getDistanceBetweenRectangleAndPoint({
      rectangle: rectangles[key],
      point
    })
    if (d < minD) {
      ret = key
      minD = d
    }
  })
  return ret
}
