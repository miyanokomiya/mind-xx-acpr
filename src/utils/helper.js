import * as dateFns from 'date-fns'

export function dateTimeFormat(date) {
  return dateFns.format(date, 'yyyy/MM/dd HH:mm:ss')
}

export function dateFormat(date) {
  return dateFns.format(date, 'yyyy/MM/dd')
}
