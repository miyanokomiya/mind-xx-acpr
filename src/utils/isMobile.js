import Vue from 'vue'
const isMobile = require('ismobilejs')

const plugin = {
  install: function(Vue) {
    Vue.util.defineReactive(Vue.prototype, '$isMobile', isMobile)
  },
}
Vue.use(plugin)
