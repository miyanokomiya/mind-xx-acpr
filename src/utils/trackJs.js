import Vue from 'vue'

if (!window.trackJs) {
  window.trackJs = {
    track () {
      console.log('trackJs', arguments)
    }
  }
}

Vue.config.errorHandler = (err, vm, info) => {
  // Log properties passed to the component if there are any
  if (vm.$options.propsData) {
    console.log('Props passed to component', vm.$options.propsData)
  }

  // Emit component name and also the lifecycle hook the error occured in if present
  var infoMessage = `Error in component: <${vm.$options.name} />\n`
  if (info) {
    infoMessage += `Error occurred during lifecycle hook: ${info}\n`
  }

  // This puts the additional error information in the Telemetry Timeline
  console.log(infoMessage)

  // Track the native JS error
  window.trackJs.track(err)
}
