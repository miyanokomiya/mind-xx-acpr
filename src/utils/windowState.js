import Vue from 'vue'

const windowStatePlugin = {
  install: function(Vue) {
    // ウィンドウの状態
    var state = {
      scrollX: 0,
      scrollY: 0,
      width: 0,
      height: 0,
    }
    // // スクロール数値を取得
    // var onScroll = function () {
    //   state.scrollX = window.pageXOffset
    //   state.scrollY = window.pageYOffset
    // }
    // document.addEventListener('scroll', onScroll)
    // ウィンドウのサイズを取得
    var onResize = function() {
      state.width = document.documentElement.clientWidth
      state.height = document.documentElement.clientHeight
    }
    window.addEventListener('resize', onResize)
    onResize()

    // // 初期値を更新
    // window.addEventListener('load', onScroll)
    // プロパティ $window を定義
    Vue.util.defineReactive(Vue.prototype, '$window', state)
  },
}
Vue.use(windowStatePlugin)
