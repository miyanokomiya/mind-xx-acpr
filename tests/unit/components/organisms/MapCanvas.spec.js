import { shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Target from '@/components/organisms/MapCanvas.vue'
import { createNode, createUser, createComment } from '@/utils/model'
import { ROOT_NODE } from '@/constants'

const vuetify = new Vuetify()

const getNodes = () => {
  return {
    [ROOT_NODE]: createNode({
      text: 'root',
      children: ['a', 'b', 'c'],
      oppositeChildren: ['zz', 'xx'],
    }),
    a: createNode({
      text: 'aaaa',
      children: ['d'],
    }),
    b: createNode({ text: 'bbbbbbb\nbbbbb' }),
    c: createNode({ text: 'cc\ncc\nccc' }),
    d: createNode({ text: 'dddd' }),
    xx: createNode({ text: 'xxxx' }),
    zz: createNode({ text: 'zzzz', children: ['yy', 'uu'] }),
    yy: createNode({ text: 'yyyy' }),
    uu: createNode({ text: 'uuuu' }),
  }
}
const getComments = () => {
  return {
    a: createComment({ uid: 'a', text: 'aaa\naaa', nodeId: 'b' }),
    aa: createComment({ uid: 'a', text: 'aaa\naaa', nodeId: 'b' }),
    aaa: createComment({ uid: 'a', text: 'aaa\naaa', nodeId: 'c' }),
    aaaa: createComment({ uid: 'a', text: 'aaa\naaa', nodeId: 'b' }),
    b: createComment({ uid: 'b', text: 'bbb', nodeId: 'c' }),
  }
}
const getUsers = () => {
  return {
    a: createUser({ displayName: 'aabb' }),
    b: createUser({ displayName: 'bbaa' }),
    user: createUser({ uid: 'user' }),
  }
}

const template = `
  <v-app>
    <div style="width: 620px; height: 600px;">
      <Target
        :width="600"
        :height="600"
        :nodes="nodes"
        :selectedNodes="selectedNodes"
        :canWrite="true"
        :user="user"
        :comments="comments"
        :users="users"
      />
    </div>
  </v-app>
`

describe('components/organisms/MapCanvas.vue', () => {
  describe('snapshot', () => {
    it('no data', () => {
      const wrapper = shallowMount(Target, {
        vuetify,
        propsData: {
          width: 100,
          hegith: 200,
          nodes: { [ROOT_NODE]: createNode() },
          selectedNodes: {},
          user: {},
          comments: {},
          users: {},
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
    it('full data', () => {
      const wrapper = shallowMount(Target, {
        vuetify,
        propsData: {
          width: 100,
          hegith: 200,
          nodes: getNodes(),
          selectedNodes: {},
          user: getUsers().user,
          comments: getComments(),
          users: getUsers(),
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
    it('can write and select nodes', () => {
      const wrapper = shallowMount(Target, {
        vuetify,
        propsData: {
          width: 100,
          hegith: 200,
          nodes: getNodes(),
          selectedNodes: { a: true, b: true },
          user: getUsers().user,
          comments: getComments(),
          users: getUsers(),
          canWrite: true,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('selectNode', () => {
      const getWrapper = () => {
        const wrapper = shallowMount(Target, {
          vuetify,
          propsData: {
            width: 100,
            hegith: 200,
            nodes: getNodes(),
            selectedNodes: { a: true, b: true },
            canWrite: true,
          },
          computed: {
            scale: {
              get() {
                return 1
              },
              set() {},
            },
          },
          methods: {
            moveViewToCoverSelectedNode: jest.fn(),
          },
        })
        wrapper.setData({
          nodePositions: { c: { x: 10, y: 20 } },
          nodeSizes: { c: { x: 100, y: 200 } },
        })
        return wrapper
      }
      it('single select', () => {
        const wrapper = getWrapper()
        wrapper.vm.selectNode('c')
        expect(wrapper.emitted('setSelectedNodes')).toEqual([[{ c: true }]])
      })
      it('multi select', () => {
        const wrapper = getWrapper()
        wrapper.vm.selectNode('c', true)
        expect(wrapper.emitted('setSelectedNodes')).toEqual([
          [{ a: true, b: true, c: true }],
        ])
      })
      it('can not write', () => {
        const wrapper = getWrapper()
        wrapper.setProps({ canWrite: false })
        wrapper.vm.selectNode('c', true)
        expect(wrapper.emitted('setSelectedNodes')).toBeFalsy()
      })
    })
  })
})
