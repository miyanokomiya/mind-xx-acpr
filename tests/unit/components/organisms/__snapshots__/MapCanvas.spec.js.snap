// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/MapCanvas.vue snapshot can write and select nodes 1`] = `
<div
  class="map-canvas-wrapper"
  hegith="200"
>
  <div
    class="canvas-wrapper"
    tabindex="1"
  >
    <!---->
     
    <svgcanvas-stub
      height="400"
      scale="1"
      width="100"
      x="0"
      y="0"
    >
         
      <svgtextrectangle-stub
        checked="-1"
        childrencount="3"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        root="true"
        stroke="black"
        strokewidth="1"
        text="root"
        textfill="#000000"
        x="-25"
        y="0"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="1"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="blue"
        strokewidth="2"
        text="aaaa"
        textfill="#000000"
        x="65"
        y="-40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="3"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="blue"
        strokewidth="2"
        text="bbbbbbb
bbbbb"
        textfill="#000000"
        x="65"
        y="0"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="2"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="cc
cc
ccc"
        textfill="#000000"
        x="65"
        y="40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="dddd"
        textfill="#000000"
        x="155"
        y="-40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="xxxx"
        textfill="#000000"
        x="-115"
        y="40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="2"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="zzzz"
        textfill="#000000"
        x="-115"
        y="-20"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="yyyy"
        textfill="#000000"
        x="-205"
        y="-40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="uuuu"
        textfill="#000000"
        x="-205"
        y="0"
      />
       
      <!---->
       
    </svgcanvas-stub>
  </div>
   
  <div
    class="scale-tool"
  >
    <scaletoolbox-stub
      max="15"
      min="-17.545081794683426"
      scalerate="0"
      step="0.1"
    />
  </div>
   
  <div
    class="history-tool"
  >
    <historytoolbox-stub />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <commentlist-stub
    class="comment-list"
    comments="[object Object]"
    style="display: none;"
    user="[object Object]"
    users="[object Object]"
  />
   
  <v-snackbar-stub
    bottom="true"
    right="true"
    timeout="2000"
  >
    
    One more click to delete.
  
  </v-snackbar-stub>
</div>
`;

exports[`components/organisms/MapCanvas.vue snapshot full data 1`] = `
<div
  class="map-canvas-wrapper"
  hegith="200"
>
  <div
    class="canvas-wrapper"
    tabindex=""
  >
    <v-icon-stub
      class="lock-button"
    >
      lock
    </v-icon-stub>
     
    <svgcanvas-stub
      height="400"
      scale="1"
      width="100"
      x="0"
      y="0"
    >
         
      <svgtextrectangle-stub
        checked="-1"
        childrencount="3"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        root="true"
        stroke="black"
        strokewidth="1"
        text="root"
        textfill="#000000"
        x="-25"
        y="0"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="1"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="aaaa"
        textfill="#000000"
        x="65"
        y="-40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="3"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="bbbbbbb
bbbbb"
        textfill="#000000"
        x="65"
        y="0"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="2"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="cc
cc
ccc"
        textfill="#000000"
        x="65"
        y="40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="dddd"
        textfill="#000000"
        x="155"
        y="-40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="xxxx"
        textfill="#000000"
        x="-115"
        y="40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="2"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="zzzz"
        textfill="#000000"
        x="-115"
        y="-20"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="yyyy"
        textfill="#000000"
        x="-205"
        y="-40"
      />
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        stroke="black"
        strokewidth="1"
        text="uuuu"
        textfill="#000000"
        x="-205"
        y="0"
      />
       
      <!---->
       
    </svgcanvas-stub>
  </div>
   
  <div
    class="scale-tool"
  >
    <scaletoolbox-stub
      max="15"
      min="-17.545081794683426"
      scalerate="0"
      step="0.1"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <commentlist-stub
    class="comment-list"
    comments="[object Object]"
    style="display: none;"
    user="[object Object]"
    users="[object Object]"
  />
   
  <v-snackbar-stub
    bottom="true"
    right="true"
    timeout="2000"
  >
    
    One more click to delete.
  
  </v-snackbar-stub>
</div>
`;

exports[`components/organisms/MapCanvas.vue snapshot no data 1`] = `
<div
  class="map-canvas-wrapper"
  hegith="200"
>
  <div
    class="canvas-wrapper"
    tabindex=""
  >
    <v-icon-stub
      class="lock-button"
    >
      lock
    </v-icon-stub>
     
    <svgcanvas-stub
      height="400"
      scale="1"
      width="100"
      x="0"
      y="0"
    >
         
      <svgtextrectangle-stub
        checked="-1"
        childrencount="0"
        class="mind-node"
        commentcount="0"
        fill="#B3E5FC"
        fontsize="10"
        hiddenfamilycount="0"
        root="true"
        stroke="black"
        strokewidth="1"
        text=""
        textfill="#000000"
        x="-25"
        y="0"
      />
       
      <!---->
       
    </svgcanvas-stub>
  </div>
   
  <div
    class="scale-tool"
  >
    <scaletoolbox-stub
      max="15"
      min="-17.545081794683426"
      scalerate="0"
      step="0.1"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <commentlist-stub
    class="comment-list"
    comments="[object Object]"
    style="display: none;"
    user="[object Object]"
    users="[object Object]"
  />
   
  <v-snackbar-stub
    bottom="true"
    right="true"
    timeout="2000"
  >
    
    One more click to delete.
  
  </v-snackbar-stub>
</div>
`;
